language-server/src/dotty/tools/languageserver/Worksheet.scala

```diff
@@ -0,0 +1,248 @@
+ *
+ * @param javaExec      The path to the `java` executable.
+ * @param userClasspath The REPL classpath
+ * @param cancelChecker The token that indicates whether evaluation has been cancelled.
+ */
+private class Evaluator private (javaExec: String,
+                                 userClasspath: String,
+                                 cancelChecker: CancelChecker) {
+  private val process =
+    new ProcessBuilder(
+      javaExec,
+      "-classpath", sys.props("java.class.path"),
+      dotty.tools.repl.WorksheetMain.getClass.getName.init,
+      "-classpath", userClasspath,
+      "-color:never")
+       .redirectErrorStream(true)
+       .start()
+
+  // The stream that we use to send commands to the REPL
+  private val processInput = new PrintStream(process.getOutputStream())
+
+  // Messages coming out of the REPL
+  private val processOutput = new ReplReader(process.getInputStream())
+  processOutput.start()
+
+  // The thread that monitors cancellation
+  private val cancellationThread = new CancellationThread(cancelChecker, this)
+  cancellationThread.start()
+
+  // Wait for the REPL to be ready
+  processOutput.next()
+
+  /** Is the process that runs the REPL still alive? */
+  def isAlive(): Boolean = process.isAlive()
+
+  /**
+   * Submit `command` to the REPL, wait for the result.
+   *
+   * @param command The command to evaluate.
+   * @return The result from the REPL.
+   */
+  def eval(command: String): Option[String] = {
+    processInput.println(command)
+    processInput.flush()
+    processOutput.next().map(_.trim)
+  }
+
+  /**
+   * Reset the REPL to its initial state, update the cancel checker.
+   */
+  def reset(cancelChecker: CancelChecker): Unit = {
+    cancellationThread.setCancelChecker(cancelChecker)
+    eval(":reset")
+  }
+
+  /** Terminate this JVM. */
+  def exit(): Unit = {
+    processOutput.interrupt()
+    process.destroyForcibly()
+    Evaluator.previousEvaluator = None
+    cancellationThread.interrupt()
+  }
+
+}
+
+/**
+ * Regularly check whether execution has been cancelled, kill REPL if it is.
+ */
+private class CancellationThread(private[this] var cancelChecker: CancelChecker,
+                                 evaluator: Evaluator) extends Thread {
+  private final val checkCancelledDelayMs = 50
+
+  override def run(): Unit = {
+    try {
+      while (evaluator.isAlive() && !Thread.interrupted()) {
+        cancelChecker.checkCanceled()
+        Thread.sleep(checkCancelledDelayMs)
+      }
+    } catch {
+      case _: CancellationException => evaluator.exit()
+      case _: InterruptedException => evaluator.exit()
+    }
+  }
+
+  def setCancelChecker(cancelChecker: CancelChecker): Unit = {
+    this.cancelChecker = cancelChecker
+  }
+}
+
+/**
+ * Reads the output from the REPL and makes it available via `next()`.
+ *
+ * @param stream The stream of messages coming out of the REPL.
+ */
+private class ReplReader(stream: InputStream) extends Thread {
```

Comment:
That's a lot of classes in one file, maybe make a worksheet/ subdirectory instead?